tools/clang/include/clang/SPIRV/EmitVisitor.h

```diff
@@ -8,20 +8,85 @@
 #ifndef LLVM_CLANG_SPIRV_EMITVISITOR_H
 #define LLVM_CLANG_SPIRV_EMITVISITOR_H
 
+#include "clang/SPIRV/SPIRVContext.h"
 #include "clang/SPIRV/SpirvVisitor.h"
+#include "llvm/ADT/DenseMap.h"
 
 namespace clang {
 namespace spirv {
 
 class SpirvModule;
 class SpirvFunction;
 class SpirvBasicBlock;
+class SpirvType;
+
+// Provides DenseMapInfo for SpirvLayoutRule so that we can use it as key to
+// DenseMap.
+//
+// Mostly from DenseMapInfo<unsigned> in DenseMapInfo.h.
+struct SpirvLayoutRuleDenseMapInfo {
+  static inline SpirvLayoutRule getEmptyKey() { return SpirvLayoutRule::Void; }
+  static inline SpirvLayoutRule getTombstoneKey() {
+    return SpirvLayoutRule::Void;
+  }
+  static unsigned getHashValue(const SpirvLayoutRule &Val) {
+    return static_cast<unsigned>(Val) * 37U;
+  }
+  static bool isEqual(const SpirvLayoutRule &LHS, const SpirvLayoutRule &RHS) {
+    return LHS == RHS;
+  }
+};
+
+class EmitTypeHandler {
+public:
+  EmitTypeHandler(SpirvContext &c, SpirvModule &m,
+                  std::vector<uint32_t> *decVec,
+                  std::vector<uint32_t> *typesVec)
+      : context(c), module(m), annotationsBinary(decVec),
+        typeConstantBinary(typesVec) {
+    assert(decVec);
+    assert(typesVec);
+  }
+
+  // Disable copy constructor/assignment.
+  EmitTypeHandler(const EmitTypeHandler &) = delete;
+  EmitTypeHandler &operator=(const EmitTypeHandler &) = delete;
+
+  // Emits the instruction for the given type into the typeConstantBinary and
+  // returns the result-id for the type.
+  uint32_t emitType(const SpirvType *, SpirvLayoutRule);
+
+  uint32_t getResultIdForType(const SpirvType *, SpirvLayoutRule,
+                              bool *alreadyExists);
+
+private:
+  void initTypeInstruction(spv::Op op);
+  void finalizeTypeInstruction();
+
+private:
+  SpirvContext &context;
+  SpirvModule &module;
+  std::vector<uint32_t> curTypeInst;
+  std::vector<uint32_t> *annotationsBinary;
+  std::vector<uint32_t> *typeConstantBinary;
+
+  // emittedTypes is a map that caches the <result-id> of types in order to
+  // avoid translating a type multiple times.
+  using LayoutRuleToTypeIdMap =
+      llvm::DenseMap<SpirvLayoutRule, uint32_t, SpirvLayoutRuleDenseMapInfo>;
+  llvm::DenseMap<const SpirvType *, LayoutRuleToTypeIdMap> emittedTypes;
+};
 
 /// \breif The visitor class that emits the SPIR-V words from the in-memory
 /// representation.
 class EmitVisitor : public Visitor {
 public:
-  EmitVisitor() = default;
+  EmitVisitor(const SpirvCodeGenOptions &opts, SpirvContext &ctx,
+              SpirvModule &mod)
+      : Visitor(opts, ctx, mod) {
+    typeHandler = new (ctx)
```

Comment:
We can just embed this type handler in `EmitVisitor` itself by having a `EmitTypeHandler` member.

Typically we only want SPIR-V IMR entities to be allocated from the SPIR-V context b/c there are a lot of them and they persist for a long time. For the rest, does not matter that much.